app/src/main/java/me/hufman/androidautoidrive/phoneui/SpotifyAuthorizationActivity.kt

```diff
@@ -0,0 +1,240 @@
+			SpotifyScope.USER_LIBRARY_READ.uri
+	)
+
+	/**
+	 * Performs the authorization request
+	 */
+	private fun doAuth() {
+		Log.d(TAG, "Authorization started")
+		try {
+			authIntentLatch.await()
+		} catch (ex: InterruptedException) {
+			Log.w(TAG, "Interrupted while waiting for auth intent")
+		}
+
+		val intent = authService.getAuthorizationRequestIntent(
+				authRequest.get(),
+				authIntent.get())
+
+		startActivityForResult(intent, REQUEST_CODE_SPOTIFY_LOGIN)
+	}
+
+	private fun warmUpBrowser() {
+		authIntentLatch = CountDownLatch(1)
+		Log.d(TAG, "Warming up browser instance for auth request")
+		val intentBuilder = authService.createCustomTabsIntentBuilder(authRequest.get().toUri())
+		authIntent.set(intentBuilder.build())
+		authIntentLatch.countDown()
+	}
+
+	private fun createAuthRequest() {
+		Log.d(TAG, "Creating auth request")
+		val codeChallenge = getSpotifyPkceCodeChallenge(CODE_VERIFIER)
+		val codeChallengeMethod = "S256"
+		val clientId = packageManager.getApplicationInfo(packageName, PackageManager.GET_META_DATA)
+				.metaData.getString("com.spotify.music.API_KEY", "unavailable")
+		val authRequestBuilder = AuthorizationRequest.Builder(
+				authStateManager.getAuthorizationServiceConfiguration()!!,
+				clientId,
+				ResponseTypeValues.CODE,
+				Uri.parse(SpotifyAppController.REDIRECT_URI))
+				.setScopes(scopes)
+				.setCodeVerifier(CODE_VERIFIER, codeChallenge, codeChallengeMethod)
+		authRequest.set(authRequestBuilder.build())
+	}
+
+	private fun initializeAppAuth() {
+		Log.d(TAG, "Initializing AppAuth")
+		recreateAuthorizationService()
+
+		if (authStateManager.getAuthorizationServiceConfiguration() == null) {
+			val authEndpointUri = Uri.parse("https://accounts.spotify.com/authorize")
+			val tokenEndpointUri = Uri.parse("https://accounts.spotify.com/api/token")
+			val config = AuthorizationServiceConfiguration(authEndpointUri, tokenEndpointUri)
+			authStateManager.replaceAuthState(AuthState(config))
+		}
+
+		runBlocking {
+			createAuthRequest()
+			warmUpBrowser()
+		}
+	}
+
+	private fun recreateAuthorizationService() {
+		if (this::authService.isInitialized) {
+			Log.d(TAG, "Discarding existing AuthService instance")
+			authService.dispose()
+		}
+		authService = createAuthorizationService()
+		authRequest.set(null)
+		authIntent.set(null)
+	}
+
+	private fun createAuthorizationService(): AuthorizationService {
+		Log.d(TAG, "Creating authorization service")
+		val builder = AppAuthConfiguration.Builder()
+		builder.setBrowserMatcher(BrowserWhitelist(
+				VersionedBrowserMatcher.CHROME_CUSTOM_TAB,
+				VersionedBrowserMatcher.SAMSUNG_CUSTOM_TAB))
+		builder.setConnectionBuilder(DefaultConnectionBuilder.INSTANCE)
+		return AuthorizationService(this, builder.build())
+	}
+
+	private fun processAuthorizationCode(requestCode: Int, resultCode: Int, data: Intent?) {
+		if (requestCode == REQUEST_CODE_SPOTIFY_LOGIN) {
+			if (resultCode == RESULT_CANCELED) {
+				onAuthorizationCancelled()
```

Comment:
The spelling of `canceled` vs `cancelled` should be consistent

Reply:
I always thought cancelled was the correct spelling, but apparently canceled is the valid American spelling, TIL!

Reply:
Ugh not sure why I kept changing the spelling back and forth throughout, fixed to be consistent and use the spelling "canceled"